app/models/concerns/cached_counting.rb

```diff
@@ -3,75 +3,158 @@
+  SLEEP_SECONDS = 1
+  FLUSH_DB_ITERATIONS = 60
+  MUTEX = Mutex.new
+
+  def self.disable
+    @enabled = false
+    if @thread && @thread.alive?
+      @thread.wakeup
+      @thread.join
     end
+  end
 
-    # auto flush if backlog is larger than this
-    self.autoflush = 2000
+  def self.enabled?
+    @enabled != false
+  end
 
-    # auto flush if older than this
-    self.autoflush_seconds = 5.minutes
+  def self.enable
+    @enabled = true
+  end
 
-    self.last_flush = Time.now.utc
+  def self.reset
+    clear_queue!
+    clear_flush_to_db_lock!
   end
 
-  class_methods do
-    def perform_increment!(key, opts = nil)
-      val = DiscourseRedis.ignore_readonly do
-        LUA_INCR_AND_EXPIRE.eval(
-          Discourse.redis.without_namespace,
-          [Discourse.redis.namespace_key(key)],
-          [EXPIRE_CACHE_AFTER]
-        ).to_i
+  def self.ensure_thread!
+    return if !enabled?
+
+    MUTEX.synchronize do
+      if !@thread&.alive?
+        @thread = nil
       end
+      @thread ||= Thread.new { thread_loop }
+    end
+  end
 
-      # readonly mode it is going to be nil, skip
-      return if val.nil?
+  def self.thread_loop
+    iterations = 0
+    while true
+      break if !enabled?
+
+      sleep SLEEP_SECONDS
+      flush_in_memory
+      if (iterations >= FLUSH_DB_ITERATIONS) || @flush
+        iterations = 0
+        flush_to_db
+        @flush = false
+      end
+      iterations += 1
+    end
+  end
 
-      autoflush = (opts && opts[:autoflush]) || self.autoflush
-      if autoflush > 0 && val >= autoflush
-        write_cache!
-        return
+  def self.flush
+    @flush = true
+    @thread.wakeup
+    while @flush
+      sleep 0.001
+    end
+  end
+
+  COUNTER_PREFIX = "__DCC__"
+
+  def self.flush_in_memory
+    counts = nil
+    while QUEUE.length > 0
+      # only 1 consumer, no need to avoid blocking
+      key, klass, db, time = QUEUE.deq
+      _redis_key = "#{COUNTER_PREFIX},#{klass},#{db},#{time.strftime("%Y%m%d")},#{key}"
+      counts ||= Hash.new(0)
+      counts[_redis_key] += 1
+    end
+
+    if counts
+      counts.each do |redis_key, count|
+        # TODO this whole loop can be done in a single LUA script
+        # concerns:
+        # - Is there a limit of params, will we need to chunk it
+        # - Would this lock up redis for too long, chunk it due to that?
+        Discourse.redis.without_namespace.incrby(redis_key, count)
       end
+    end
+  end
 
-      if (Time.now.utc - last_flush).to_i > autoflush_seconds
-        write_cache!
+  DB_FLUSH_COOLDOWN_SECONDS = 60
+  DB_COOLDOWN_KEY = "cached_counting_cooldown"
+
+  def self.flush_to_db
+    redis = Discourse.redis.without_namespace
+    DistributedMutex.synchronize("flush_counters_to_db", redis: redis, validity: 5.minutes) do
+      if allowed_to_flush_to_db?
+        # TODO can be done in a single eval (including keys call)
+        # same concern as above
+        redis.keys("#{COUNTER_PREFIX}*").each do |key|
```

Comment:
`KEYS` can be quite risky - the docs say not to use it in production because it blocks the whole database.

Given that we're no longer relying on Redis to handle key expiration... I wonder if we should move all the values for a given `COUNTER_PREFIX` into a single redis `hash`. We can use `HINCRBY` to increment values and `HGETALL` to fetch all values for flushing to the db. 🤔